contracts/Aavegotchi/facets/PeripheryFacet.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.1;
+
+import "../libraries/LibItems.sol";
+import "../libraries/LibAppStorage.sol";
+import "../libraries/LibERC1155Marketplace.sol";
+import "../../shared/libraries/LibStrings.sol";
+
+contract PeripheryFacet is Modifiers {
+    ///READ FUNCTIONS
+
+    function peripheryBalanceOf(address _owner, uint256 _id) external view returns (uint256 balance_) {
+        balance_ = s.ownerItemBalances[_owner][_id];
+    }
+
+    function peripheryBalanceOfBatch(address[] calldata _owners, uint256[] calldata _ids) external view returns (uint256[] memory bals_) {
+        require(_owners.length == _ids.length, "ItemsFacet: _owners length not same as _ids length");
+        bals_ = new uint256[](_owners.length);
+        for (uint256 i; i < _owners.length; i++) {
+            uint256 id = _ids[i];
+            address owner = _owners[i];
+            bals_[i] = s.ownerItemBalances[owner][id];
+        }
+    }
+
+    function peripheryUri(uint256 _id) external view returns (string memory) {
+        require(_id < s.itemTypes.length, "ItemsFacet: Item _id not found");
+        return LibStrings.strWithUint(s.itemsBaseUri, _id);
+    }
+
+    function peripheryIsApprovedForAll(address _owner, address _operator) external view returns (bool approved_) {
+        approved_ = s.operators[_owner][_operator];
+    }
+
+    //WRITE
+
+    function peripherySetApprovalForAll(
+        address _operator,
+        bool _approved,
+        address _onBehalfOf
+    ) external onlyPeriphery {
+        require(tx.origin == _onBehalfOf, "PeripheryFacet: Approval Failed");
+        address sender = tx.origin;
```

Comment:
in regards to https://github.com/aavegotchi/aavegotchi-contracts/pull/150/files/49db6c54892a2ad6e1e64c0c6f01aafbe28d8b3f#r984912196

@pakim249 you can approve am arbitrary address here

Reply:
The problem i am pointing out is not setting approvals.

Take for example setting an approval to an opensea contract.

An EOA will have no problem setting an operator as the opensea contract.

The problem lies afterwards, where the opensea contract calls safeTransferFrom. In this case, tx.origin will NOT be the opensea contract address, because tx.origin is always the EOA that originates the transaction. As such, approving a smart contract address is not possible in this implementation.

Reply:
oh..so in short, because EOAs cannot be contracts, smart contracts cannot approve other addresses

Reply:
Yes, and contracts cannot use safeTransferFrom with an approval. In general, usage of tx.origin breaks compatibility with smart contracts so I prefer never using it if possible.

Reply:
>Yes, and contracts cannot use safeTransferFrom with an approval

I don't follow here... Sorry for wasting your time😢

Reply:
As an example, let's say that we want to approve opensea to transfer my items on my behalf when someone bids on it.

Define the following:

address originalOwner; // EOA
address openSea; // contract
address bidder; // EOA

To do the buy process, originalOwner should approve openSea to transfer tokens on their behalf by calling:

setApprovalForAll(opensea, true);

Then the bidder would call the opensea contract, transferring tokens to it. At this point, the openSea contract would have to call:

safeTransferFrom(originalOwner, bidder, id, 1, "0x"); // from, to, id, value, data

This call will not succeed in this implementation. The reason is because in the peripherySafeTransferFrom sets the sender to tx.origin. In this case, tx.origin is bidder.

So the only way for the bidding process to succeed is for the originalOwner to approve the bidder's address, which requires knowing the bidder in advance. This to me does not seem like a desired behavior.

Reply:
gotcha

>i think it makes sense then to pass `msg.sender` in from the periphery

if this is done for `safeTransferFrom` , should mitigate it


sender would be the address the periphery passes in 

and the periphery would make sure that `msg.sender` is passed passed to `peripherySafeTransferFrom`

Reply:
Thanks!